feat(Add TS version): generate new image of 10.3.1 with Typescript

jbarlin opened the pull request:
Also changed the README to reflect the 10.3.1 image base that's listed in the Dockerfile

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cypress-io/cypress-docker-images?pullRequest=718) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cypress-io/cypress-docker-images?pullRequest=718) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cypress-io/cypress-docker-images?pullRequest=718) it.</sub>

admah commented:
@jbarlin thanks for submitting this. You'll also need to update the image and relevant values in `circle.yml`. Typically our automated process kicks this off, but I'm assuming you manually built this image?

jbarlin commented:
Hola!

I did build it locally and test it yes, but I just followed the guide in the CONTRIBUTING.md guide for adding a new base image, and then went back and renamed wherever it had the browser versions at the end of the image name and replaced it with "-typescript".

Looking at the `generate-included-image.js`, it makes a call to `generate-config.js`, but when I run it I get the following error:

<details>
  <summary>Click to open error</summary>
```
[jbarlin@Joshua-MBP cypress-docker-images]$ node ./scripts/generate-config.js included 10.3.1-typescript
/media/misc/github/cypress-docker-images/scripts/generate-config.js:28
const sanitizedImageType = getImageType(image)
                                        ^

ReferenceError: Cannot access 'image' before initialization
    at Object.<anonymous> (/media/misc/github/cypress-docker-images/scripts/generate-config.js:28:41)
    at Module._compile (node:internal/modules/cjs/loader:1120:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1174:10)
    at Module.load (node:internal/modules/cjs/loader:998:32)
    at Module._load (node:internal/modules/cjs/loader:839:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47

Node.js v18.6.0

```
</details>

I fixed this reference and another two that weren't working, re-ran the script and it seems to have generated correctly - let me know if you need to me to change the script back/fix the config 😄

jbarlin commented:
`yarn test` looks like it's trying to test the config building script... that I've just altered to not yield the error it was giving.

That said, I don't think what I have changed about the `generate-config` script should cause that much of a material difference, unless I've broken it in some other way?

admah commented:
@jbarlin have you rebased off of `master` recently?

jbarlin commented:
It's forked off master - commit 3be11cfec32052c3e8c3836c587b7288ffa8321c

GitHub tells me that there is no new commits to fetch, and it looks like that's the case

jbarlin commented:
Looking at the master branch pipeline, the CI marked it as passing but "lint" did get the same error I was - https://app.circleci.com/pipelines/github/cypress-io/cypress-docker-images/1006/workflows/09d5a982-48d4-4166-9874-dc6171b9da32/jobs/37781

jbarlin commented:
Scratch that, a different error. Still, it threw an error and then marked it as passing.

admah commented:
@jbarlin to fix the `lint` error, you should be able to delete `scripts/__tests__/__snapshots__/generate-config-snapshots.test.js.snap` and then run `yarn test` again to generate new snapshots. There was a commit where those were incorrect.

admah commented:
@jbarlin you should be get a new snapshot file to commit if you run `yarn test` - not just delete the existing one.

jbarlin commented:
@admah lint is still getting the same errors on CI: https://app.circleci.com/pipelines/github/cypress-io/cypress-docker-images/1009/workflows/96a6a4a6-9647-4c0e-a8c5-1207878cb165/jobs/37798?invite=true#step-108-16
It would be a node version issue by the looks of things. Changing node version on my local machine I can recreate the error it is getting. I also still get the same error in `generate-config` on the older node versions - am I right to leave those changes in?

admah commented:
@jbarlin I pulled down those changes and ran them locally. Everything looked good to me and tests passed as expected.

admah commented on circle.yml:
```diff
@@ -411,15 +411,15 @@ jobs:
 
                     ## see https://docs.docker.com/desktop/multi-arch/
-                    echo "$DOCKERHUB_PASS" | docker login -u "$DOCKERHUB_USERNAME" --password-stdin
+                    docker login -u "$DOCKERHUB_USERNAME" --password "$DOCKERHUB_PASS"
```
These changes were already present in #717 so not sure why they're showing up here. I'm just adding this comment as a note.

jbarlin commented:
@admah awesome! Want me to squash the commits?

admah commented on scripts/generate-config.js:
```diff
@@ -23,9 +23,9 @@ const splitImageFolderName = (folderName) => {
 }
 
 const getImageType = (image) => {
```
```suggestion
const getImageType = (imageType) => {
```

admah commented on scripts/generate-config.js:
```diff
@@ -23,9 +23,9 @@ const splitImageFolderName = (folderName) => {
 
 const getImageType = (image) => {
-  return image.name.includes('base') ? 'base' : image.name.includes('browser') ? 'browser' : 'included'
+  return image.includes('base') ? 'base' : image.includes('browser') ? 'browser' : 'included'
```
```suggestion
  return imageType.includes('base') ? 'base' : imageType.includes('browser') ? 'browser' : 'included'
```

I think it's better to keep this name in parity with what you have on line 28.

admah reviewed (changes requested):
I had a couple small suggestions. Once those are addressed, this is good to go. Thank you!

jbarlin commented:
@admah Done!

admah commented:
@jbarlin thanks for all of your work on this PR 🎉

jbarlin commented:
@admah Thank you!
